

<h4>Changelog</h4>

Here you can find the recent changes made to the patchwork and patchworkCG packages.
<br />

<!-- <hr class="alt1" />

<h5 style="display:inline-block;">patchwork v2.4 - </h5><div style="display:inline-block; width:1px;"></div>
<h6 style="display:inline-block;">2013</h6>
<ul>
	<li>Moved Pysam out of package due to auto-install issues. Now you will have to manually install it.
	See Patchwork -> Installation </li>
</ul> -->

 <hr class="alt1" />

<h5 style="display:inline-block;">TAPS 2.1 - </h5><div style="display:inline-block; width:1px;"></div>
<h6 style="display:inline-block;">04/03/2014</h6>
<ul>
	<li>New function: TAPS_estimate(). Automatic estimation of allele-specific copynumbers for all samples that have completed TAPS_plot()</li>
	<li>New function: TAPS_click(). Use a GUI to accept TAPS_estimates() values or add your own interpretation through direct interaction with the TAPS_plot() plots. </li>
	<li>New function: TAPS_region(). Zoom in on a region specified by user which includes known gene information.</li>
	<li>Updates to homepage regarding TAPS usage.</li>
</ul>

<hr class="alt1" />

<h5 style="display:inline-block;">patchwork v2.4 & TAPS 2.0 - </h5><div style="display:inline-block; width:1px;"></div>
<h6 style="display:inline-block;">11/09/2013</h6>
<ul>
	<li>Complete makeover of patchwork plots visualization. Now includes cytoband information. </li>
	<li>New function: patchwork.region(). Shows a region you specify of a chromosome. Contains Known Gene information </li>
	<li>New function: TAPS_compare(). Compare aberration frequencies in two groups of samples.</li>
	<li>New function: TAPS_freq(). Summary of aberration frequencies in several samples.</li>
</ul>


 <hr class="alt1" />

<h5 style="display:inline-block;">patchwork v2.3 - </h5><div style="display:inline-block; width:1px;"></div>
<h6 style="display:inline-block;">16/08/2013</h6>
<ul>
	<li>Moved Pysam out of package due to auto-install issues. Now you will have to manually install it.
	See Patchwork -> Installation </li>
</ul>

<hr class="alt1" />

<h5 style="display:inline-block;">patchwork & TAPS - </h5><div style="display:inline-block; width:1px;"></div>
<h6 style="display:inline-block;">18/06/2013</h6>
<ul>
	<li>Buggfix for Patchwork</li>
	<li>Buggfixes for TAPS</li>
	<li>Added information regarding the tool TAPS to homepage</li>
	<li>Patchwork now gives you an additional file, &lt;sample_name&gt;_somatic.Rdata, which contains possible, unfiltered, somatic variants.</li>
</ul>

<hr class="alt1" />

<h5 style="display:inline-block;">Homepage - </h5><div style="display:inline-block; width:1px;"></div>
<h6 style="display:inline-block;">11/09/2012</h6>
<ul>
	<li>Changes and updates to all information in tabs under Patchwork</li>
</ul>

<hr class="alt1" />

<h5 style="display:inline-block;">patchwork v2.2 - </h5><div style="display:inline-block; width:1px;"></div>
<h6 style="display:inline-block;">31/08/2012</h6>
<ul>
	<li>Parameter naming changes and updates</li>
	<li>Implemented mpileup compatability so that latest versions of SAMtools can be used. Note that it is still more efficient to use old pileup command</li>
	<li>Fixed an issue we were seeing with GC normalization when HG18 was used</li>
	<li>Fixed an issue with some plots being out of bounds when HG18 was used</li>
</ul>

<hr class="alt1" />

<h5 style="display:inline-block;">patchwork v2.1 - </h5><div style="display:inline-block; width:1px;"></div>
<h6 style="display:inline-block;">25/06/2012</h6>
<ul>
	<li>Improved sex-chromosome handling</li>
	<li>Added support and documentation for HG18</li>
	<li>Updated naming of files created from running patchwork</li>
</ul>


<hr class="alt1" />

<h5 style="display:inline-block;">patchworkCG v2.0 - </h5><div style="display:inline-block; width:1px;"></div>
<h6 style="display:inline-block;">17/04/2012</h6>
<ul>
	<li>Improved segmentation for Allelic Imbalance</li>
	<li>Changes to onLoad message handling</li>
</ul>

<hr class="alt1" />

No older changes have been recorded.

<br /><br /><div style="text-align:center;"><a href="#top" style="text-decoration:none;">Top</a></div>


